[SPARK-21087] [ML] CrossValidator, TrainValidationSplit expose sub models after fitting: Scala

WeichenXu123 opened the pull request:
## What changes were proposed in this pull request?

We add a parameter whether to collect the full model list when CrossValidator/TrainValidationSplit training (Default is NOT), avoid the change cause OOM)

- Add a method in CrossValidatorModel/TrainValidationSplitModel, allow user to get the model list

- CrossValidatorModelWriter add a “option”, allow user to control whether to persist the model list to disk (will persist by default).

- Note: when persisting the model list, use indices as the sub-model path

## How was this patch tested?

Test cases added.

WeichenXu123 committed:
init pr

SparkQA commented:
**[Test build #81685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81685/testReport)** for PR 19208 at commit [`46d3ab3`](https://github.com/apache/spark/commit/46d3ab3899c196311368b3383338b3d4e6d5aeaa).

SparkQA commented:
**[Test build #81685 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81685/testReport)** for PR 19208 at commit [`46d3ab3`](https://github.com/apache/spark/commit/46d3ab3899c196311368b3383338b3d4e6d5aeaa).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81685/
Test FAILed.

WeichenXu123 commented:
cc @jkbradley

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -261,17 +290,40 @@ class CrossValidatorModel private[ml] (
+  def save(path: String, persistSubModels: Boolean): Unit = {
+    write.asInstanceOf[CrossValidatorModel.CrossValidatorModelWriter]
+      .persistSubModels(persistSubModels).save(path)
+  }
```
I add this method because the `CrossValidatorModelWriter` is private. User cannot use it. But I don't know whether there is better solution.

WeichenXu123 replied:
I tried `model.write.asInstanceOf[CrossValidatorModel.CrossValidatorModelWriter]` but cannot pass complier, it is inaccessible.
Do you have some other ways ?

WeichenXu123 replied:
Discussion: Another way I think is adding an interface `def option(key: String, value: String)` into Writer. cc @jkbradley

jkbradley replied:
I agree with the last suggestion of adding ```def option(key: String, value: String)``` to mimic the SQL datasource API.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/ValidatorParams.scala:
```diff
@@ -150,20 +150,14 @@ private[ml] object ValidatorParams {
+      .map { case ParamPair(p, v) =>
+        p.name -> parse(p.jsonEncode(v))
+      }.toList ++ List("estimatorParamMaps" -> parse(estimatorParamMapsJson))
+    )
```
Improve code here. So that we don't need to add code for each parameter. Now we have 3 new added parameter: (parallelism, collectSubModels, persistSubModelPath), all added only in CV/TVS estimator. The old code here is easy to cause bugs if we forgot to update it when we add new params.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -212,14 +238,12 @@ object CrossValidator extends MLReadable[CrossValidator] {
         .setEstimatorParamMaps(estimatorParamMaps)
-        .setNumFolds(numFolds)
-        .setSeed(seed)
+      DefaultParamsReader.getAndSetParams(cv, metadata, skipParams = List("estimatorParamMaps"))
```
Use `getAndSetParams` instead of setting all params manually. This simplify code, and it can keep read/write compatibility.

WeichenXu123 committed:
fix style

SparkQA commented:
**[Test build #81686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81686/testReport)** for PR 19208 at commit [`ae13440`](https://github.com/apache/spark/commit/ae13440fd2220e28b58df52836f55fe5ed77c43f).

SparkQA commented:
**[Test build #81686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81686/testReport)** for PR 19208 at commit [`ae13440`](https://github.com/apache/spark/commit/ae13440fd2220e28b58df52836f55fe5ed77c43f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81686/
Test PASSed.

jkbradley commented:
CC @hhbyyh and @MLnick  Does this look reasonable to you?

And @hhbyyh would you want to split off a new JIRA for your original solution of adding an option to dump models to disk?  Then we could revive your PR.

WeichenXu123 commented:
oh...sorry for that, I integrate @hhbyyh's old PR into this new one, because I found the code "dump models to disk" and "collect models" seem to be cohesive and split them will cause some conflicts when merging. @jkbradley

jkbradley commented:
Synced offline: I hadn't looked carefully and seen the 2 issues had been merged.  @WeichenXu123 said he will split the work in 2, adding one parameter first.

WeichenXu123 committed:
remove code for dump models to disk

WeichenXu123 commented:
@jkbradley I split this PR, removed the code for "dump models to disk", so the PR will be smaller and easier to review. When this PR merged, I will create follow-up PR for "dump models to disk". Thanks!

SparkQA commented:
**[Test build #81767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81767/testReport)** for PR 19208 at commit [`e0f4ce6`](https://github.com/apache/spark/commit/e0f4ce6b21e5179e10a4a8640a3dd1aa0038e291).

SparkQA commented:
**[Test build #81767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81767/testReport)** for PR 19208 at commit [`e0f4ce6`](https://github.com/apache/spark/commit/e0f4ce6b21e5179e10a4a8640a3dd1aa0038e291).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81767/
Test FAILed.

WeichenXu123 commented:
Jenkins, test this please.

SparkQA commented:
**[Test build #81772 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81772/testReport)** for PR 19208 at commit [`e0f4ce6`](https://github.com/apache/spark/commit/e0f4ce6b21e5179e10a4a8640a3dd1aa0038e291).

SparkQA commented:
**[Test build #81772 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81772/testReport)** for PR 19208 at commit [`e0f4ce6`](https://github.com/apache/spark/commit/e0f4ce6b21e5179e10a4a8640a3dd1aa0038e291).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81772/
Test PASSed.

hhbyyh commented:
It's OK to me to include the "dump model to disk" https://github.com/apache/spark/pull/18313 in this or other PR (or not).

After reading the discussion, I feel it's an overkill to support a feature like this in two ways (keeping in memory and dumping to disk). Allowing user to register a custom action after each batch of `est.fit(trainingDataset, epm)` looks like a more general solution to me, in there user may dump models to disk, collect it for later use, or evaluate with other metric. 

If you want to stick to this way which I'm not a fan of, I would only suggest to add the logic to estimate the memory the models will cost and stops the application if OOM is foreseeable.

smurching commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -276,12 +315,32 @@ object TrainValidationSplitModel extends MLReadable[TrainValidationSplitModel] {
+        val subModelsPath = new Path(path, "subModels")
+        for (paramIndex <- 0 until instance.getEstimatorParamMaps.length) {
+          val modelPath = new Path(subModelsPath, paramIndex.toString).toString
+          instance.subModels(paramIndex).asInstanceOf[MLWritable].save(modelPath)
```
Should we clean up/remove the partially-persisted `subModels` if any of these `save()` calls fail? E.g. let's say we have four subModels and the first three `save()` calls succeed but the fourth fails - should we delete the folders for the first three submodels?

smurching replied:
@WeichenXu123 Actually I don't think we have to worry about this; Pipeline persistence doesn't clean up if a stage fails to persist (see [Pipeline.scala](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala#L242))

WeichenXu123 replied:
Ah, its a good point. But currently model saving code do not have some exception handling code. e.g, overwrite saving, when save failed, it do not recover the old directory.
I think these things can be done in separated PRs.
cc @jkbradley What' your opinion ?

jkbradley replied:
Good question about cleaning up partially saved models.  I agree it'd be nice to do in the future, rather than now.

smurching reviewed (commented):
Thanks for the work, just a couple of comments!

WeichenXu123 commented:
@smurching Thanks! I will update later. And note that I will separate part of this PR to a new PR (the separated part will be a bugfix for #16774 )

WeichenXu123 commented:
I will update this PR after #19350 get merged. We need to address another issue first. Thanks!

WeichenXu123 committed:
merge master & resolve conflicts

SparkQA commented:
**[Test build #82233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82233/testReport)** for PR 19208 at commit [`77d05f6`](https://github.com/apache/spark/commit/77d05f66ab7d764a511189960af9601b41517bfd).

SparkQA commented:
**[Test build #82233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82233/testReport)** for PR 19208 at commit [`77d05f6`](https://github.com/apache/spark/commit/77d05f66ab7d764a511189960af9601b41517bfd).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82233/
Test FAILed.

WeichenXu123 committed:
address comment issues

SparkQA commented:
**[Test build #82246 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82246/testReport)** for PR 19208 at commit [`e009ee1`](https://github.com/apache/spark/commit/e009ee1145930a02c71db85c967a49f9fd7509e5).

WeichenXu123 commented:
cc @smurching code updated, thanks!

SparkQA commented:
**[Test build #82246 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82246/testReport)** for PR 19208 at commit [`e009ee1`](https://github.com/apache/spark/commit/e009ee1145930a02c71db85c967a49f9fd7509e5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82246/
Test PASSed.

SparkQA commented:
**[Test build #3972 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3972/testReport)** for PR 19208 at commit [`e009ee1`](https://github.com/apache/spark/commit/e009ee1145930a02c71db85c967a49f9fd7509e5).

jkbradley commented:
taking a look...

SparkQA commented:
**[Test build #3972 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3972/testReport)** for PR 19208 at commit [`e009ee1`](https://github.com/apache/spark/commit/e009ee1145930a02c71db85c967a49f9fd7509e5).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -117,6 +123,12 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
 
+    val collectSubModelsParam = $(collectSubModels)
+
+    var subModels: Option[Array[Array[Model[_]]]] = if (collectSubModelsParam) {
```
so this var seems unnecessary, could we just it seems like we'd be better by just collecting modelFutures in copy values (then we can avoid the mutation on L145)

WeichenXu123 replied:
@holdenk @jkbradley I already thought about this issue. The reason I use this way is:
1) When `$(collectSubModels) == false`, the `modelFutures` and `foldMetricFutures` will be executed in pipelined way, this will make sure that the `model` generated in `modelFutures` will be released in time, so that the maximum memory cost will be `numParallelism * sizeof(model)`.  If we use the way of "collecting modelFutures", it will increase the memory cost to be `$(estimatorParamMaps).length * sizeof(model)` . This is a serious issue which is discussed before.
2) IMO the mutation on L145 won't influence performance. and it do not need something like lock, there is no race condition.

holdenk replied:
I don't follow with #1, if we keep all the models (e.g. set `collectSubModelsParam`)  then the maximum memory cost will be `$(estimatorParamMaps).length * sizeof(model)` in either case? If we don't keep the models (e.g. set `collectSubModelsParam` to false) then you don't have to collect the future back at the end and there is no additional overhead.

For #2, It's not that mutation impacts performance, its that it makes the code less easy to reason about for no gain (unless I've misunderstood something about part 1).

WeichenXu123 replied:
@holdenk Oh, sorry for confusing you. Yes, if set `collectSubModelsParam` the memory cost will always be `$(estimatorParamMaps).length * sizeof(model)`. According to your suggestion, we have to duplicate code logic (but if i am wrong correct me):
- When set `collectSubModelsParam`, we cannot pipeline `modelFutures` and `foldMetricFutures`, we should execute `modelFutures` and collect results first, and modify `foldMetricFutures` logic (change it into the way passing `model` param, not by `modelFuture.map { model => ...} ).
- When not set  `collectSubModelsParam`, just keep current `modelFutures` & `foldMetricFutures` and pipeline them to execute.
So, according to your suggestion, it seems need more code. So do you still prefer this way ? Or do you have better way to implement that ?

jkbradley replied:
Sorry I didn't follow up on this before.  I think that @WeichenXu123 's argument is valid, but please say if there are issues I'm missing @holdenk

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -187,6 +191,50 @@ class CrossValidatorSuite
+      .build()
+    val eval = new BinaryClassificationEvaluator
+    val numFolds = 3
+    val subdirName = Identifiable.randomUID("testSubModels")
```
Do we need the random UID?  tempDir is already randomized.

jkbradley reviewed (commented):
I'm sending some initial comments.  One more question: What do you think about saving the submodels (if they exist) by default, rather than throwing them out by default?  If someone has explicitly set collectSubModels=true, then it's likely they would want to persist the submodels too.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -236,12 +252,17 @@ object CrossValidator extends MLReadable[CrossValidator] {
     @Since("1.2.0") val bestModel: Model[_],
-    @Since("1.5.0") val avgMetrics: Array[Double])
+    @Since("1.5.0") val avgMetrics: Array[Double],
+    @Since("2.3.0") val subModels: Option[Array[Array[Model[_]]]])
```
So also need a `setSubModels` method for setting it.

WeichenXu123 committed:
address issues from comments

WeichenXu123 commented:
ping @jkbradley Comments all addressed! Pls take a look again. Thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83388/
Test FAILed.

WeichenXu123 committed:
fix style

SparkQA commented:
**[Test build #83393 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83393/testReport)** for PR 19208 at commit [`2a83fb5`](https://github.com/apache/spark/commit/2a83fb5071aa879dbae7791eb7c956793246f8df).

SparkQA commented:
**[Test build #83393 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83393/testReport)** for PR 19208 at commit [`2a83fb5`](https://github.com/apache/spark/commit/2a83fb5071aa879dbae7791eb7c956793246f8df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83393/
Test PASSed.

jkbradley reviewed (commented):
I'm sending some comments, but I'm not done yet.

One issue is that users will have a hard time discovered the persistSubModels option.  I'd recommend we do the following:
* Make the CrossValidatorModelWriter (and TVS writer) public, and add Scala doc to them to describe the option.
* Override the write method in CrossValidatorModel so it return type CrossValidatorModelWriter (rather than a generic MLWriter).  That should make it a little easier for users to find the writer option.
* Add a note in the setCollectSubModels method about persistSubModels (to help with discoverability).

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -187,6 +191,50 @@ class CrossValidatorSuite
+
+    val savingPathWithoutSubModels = new File(subPath, "cvModel2").getPath
+    cvModel.write.option("persistSubModels", "false").save(savingPathWithoutSubModels)
+    val cvModel2 = CrossValidatorModel.load(savingPathWithoutSubModels)
```
Shall we try saving cvModel2 with persistSubModels = true and check for an Exception?

jkbradley commented:
Done with review.  I mainly review CrossValidator since some comments will apply to TrainValidationSplit as well.  Thanks for the PR!

WeichenXu123 committed:
address comments from joseph

SparkQA commented:
**[Test build #83469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83469/testReport)** for PR 19208 at commit [`f2ef609`](https://github.com/apache/spark/commit/f2ef609ab462f1c5ce3044f2b3a3e5389a77949e).

SparkQA commented:
**[Test build #83469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83469/testReport)** for PR 19208 at commit [`f2ef609`](https://github.com/apache/spark/commit/f2ef609ab462f1c5ce3044f2b3a3e5389a77949e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83469/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -262,15 +273,26 @@ class CrossValidatorModel private[ml] (
+  /**
+   * @return submodels represented in two dimension array. The index of outer array is the
+   *         fold index, and the index of inner array corresponds to the ordering of
+   *         estimatorParamsMaps
```
typo: estimatorParamMaps

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -262,15 +273,26 @@ class CrossValidatorModel private[ml] (
+   *         fold index, and the index of inner array corresponds to the ordering of
+   *         estimatorParamsMaps
+   *
+   * Note: If submodels not available, exception will be thrown. only when we set collectSubModels
```
reword, and use ```@throws``` scaladoc:
```
@throws IllegalArgumentException  if subModels are not available.  To retrieve subModels, make sure to set collectSubModels to true before fitting.
```
(Please fix wording in the error message too.)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -323,39 +338,40 @@ object CrossValidatorModel extends MLReadable[CrossValidatorModel] {
+   * "persistSubModels" will cause exception.
+   */
+  @Since("2.3.0")
   class CrossValidatorModelWriter(instance: CrossValidatorModel) extends MLWriter {
```
Although we're making this public, let's not make all of its APIs public.  Can you please make the constructor private and make this class final?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -323,39 +338,40 @@ object CrossValidatorModel extends MLReadable[CrossValidatorModel] {
+   * @param instance CrossValidatorModel instance used to construct the writer
+   *
+   * Options:
+   * CrossValidatorModelWriter support an option "persistSubModels", available value is
```
Fix wording:

CrossValidatorModelWriter supports an option "persistSubModels", with possible values "true" or "false". If you set the collectSubModels Param before fitting, then you can set "persistSubModels" to "true" in order to persist the submodels. By default, "persistSubModels" will be "true" when submodels are available and "false" otherwise.  If submodels are not available, then setting "persistSubModels" to "true" will cause an exception.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala:
```diff
@@ -108,11 +111,19 @@ abstract class MLWriter extends BaseReadWrite with Logging {
   /**
-   * `option()` handles extra options. If subclasses need to support extra options, override this
-   * method.
+   * Map store extra options for this writer.
```
"Map to store"

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala:
```diff
@@ -108,11 +111,19 @@ abstract class MLWriter extends BaseReadWrite with Logging {
+  protected val optionMap: mutable.Map[String, String] = new mutable.HashMap[String, String]()
+
+  /**
+   * `option()` handles extra options.
```
"Adds an option to the underlying MLWriter.  See the documentation for the specific model's writer for possible options.  The option name (key) is case-insensitive."

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -323,39 +338,40 @@ object CrossValidatorModel extends MLReadable[CrossValidatorModel] {
-    }
-
     override protected def saveImpl(path: String): Unit = {
+      val persistSubModels = optionMap.getOrElse("persistsubmodels",
```
Please update this so that, when the valid is not convertible to a Boolean, the user sees an error message which states the invalid value and the possible valid values.

jkbradley reviewed (commented):
Some more small comments, thanks!

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -101,6 +101,20 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
+   * Note: If set this param, when you save the returned model, you can set an option
+   * "persistSubModels" to be "true" before saving, in order to save these submodels.
+   * You can check documents of
+   * {@link org.apache.spark.ml.tuning.CrossValidatorModel.CrossValidatorModelWriter}
```
I haven't checked through TrainValidationSplit yet, but please do make sure updates to CrossValidator get applied here (and that the updates are checked for copy errors like this line).  Thanks!

WeichenXu123 replied:
I have done some search to make sure everywhere is checked.

WeichenXu123 committed:
address comments from joseph

SparkQA commented:
**[Test build #83530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83530/testReport)** for PR 19208 at commit [`7bacfca`](https://github.com/apache/spark/commit/7bacfcac2e20552bb4557614ba477cf776bdf8af).

SparkQA commented:
**[Test build #83530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83530/testReport)** for PR 19208 at commit [`7bacfca`](https://github.com/apache/spark/commit/7bacfcac2e20552bb4557614ba477cf776bdf8af).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83530/
Test FAILed.

WeichenXu123 committed:
fix mima

SparkQA commented:
**[Test build #83534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83534/testReport)** for PR 19208 at commit [`654e4d5`](https://github.com/apache/spark/commit/654e4d580889dcd2fcf7c0bea2060349190faaac).

SparkQA commented:
**[Test build #83534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83534/testReport)** for PR 19208 at commit [`654e4d5`](https://github.com/apache/spark/commit/654e4d580889dcd2fcf7c0bea2060349190faaac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83534/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -265,23 +317,58 @@ class TrainValidationSplitModel private[ml] (
+   * If subModels are not available, then setting "persistSubModels" to "true" will cause
+   * an exception.
+   */
+  final class TrainValidationSplitModelWriter private[tuning] (
```
Since annotation

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -177,7 +202,9 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
 
   @Since("2.0.0")
-  override def write: MLWriter = new TrainValidationSplit.TrainValidationSplitWriter(this)
+  override def write: TrainValidationSplit.TrainValidationSplitWriter = {
```
Was this meant to be for TrainValidationSplitModel, not the Estimator?

WeichenXu123 replied:
Ah, there is two `write` method, one for `Estimator` and another for `Model`.
We only need to change the return type of `write` method for `model`.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:
```diff
@@ -187,6 +191,55 @@ class CrossValidatorSuite
+    val eval = new BinaryClassificationEvaluator
+    val numFolds = 3
+    val subPath = new File(tempDir, "testCrossValidatorSubModels")
+    val persistSubModelsPath = new File(subPath, "subModels").toString
```
not used

jkbradley reviewed (commented):
Just a few tiny items left, thanks!

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:
```diff
@@ -301,11 +395,29 @@ object CrossValidatorModel extends MLReadable[CrossValidatorModel] {
       val bestModelPath = new Path(path, "bestModel").toString
       val bestModel = DefaultParamsReader.loadParamsInstance[Model[_]](bestModelPath, sc)
       val avgMetrics = (metadata.metadata \ "avgMetrics").extract[Seq[Double]].toArray
+      val shouldPersistSubModels = (metadata.metadata \ "persistSubModels").extract[Boolean]
```
I realized this will not be backwards compatible.  Let's make this persistSubModels optional so that we assume it is false if it is not in the metadata.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:
```diff
@@ -298,8 +385,21 @@ object TrainValidationSplitModel extends MLReadable[TrainValidationSplitModel] {
       val bestModelPath = new Path(path, "bestModel").toString
       val bestModel = DefaultParamsReader.loadParamsInstance[Model[_]](bestModelPath, sc)
       val validationMetrics = (metadata.metadata \ "validationMetrics").extract[Seq[Double]].toArray
+      val shouldPersistSubModels = (metadata.metadata \ "persistSubModels").extract[Boolean]
```
Same here; let's make this optional

jkbradley reviewed (commented):
2 more comments about backwards compatibility.  Would you mind testing this manually, saving a model from spark 2.2 and then loading it with a build of this PR?

WeichenXu123 committed:
address minor issues

SparkQA commented:
**[Test build #83823 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83823/testReport)** for PR 19208 at commit [`2bb6835`](https://github.com/apache/spark/commit/2bb683546fe4ce4008b9a20ea61a775b98548b29).

SparkQA commented:
**[Test build #83823 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83823/testReport)** for PR 19208 at commit [`2bb6835`](https://github.com/apache/spark/commit/2bb683546fe4ce4008b9a20ea61a775b98548b29).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83823/
Test FAILed.

SparkQA commented:
**[Test build #83824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83824/testReport)** for PR 19208 at commit [`7e997da`](https://github.com/apache/spark/commit/7e997da44157a9807de9c8fe8e7d2e5b66b6bfb1).

WeichenXu123 commented:
I manually tested backwards compatibility and it works fine. I paste the test code for `CrossValidator` here.

Run following code in spark-2.2 shell first:
```
import java.io.File
import org.apache.spark.ml.tuning._
import org.apache.spark.ml.{Estimator, Model, Pipeline}
import org.apache.spark.ml.classification.{LogisticRegression, LogisticRegressionModel, OneVsRest}
import org.apache.spark.ml.feature.HashingTF
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.regression.LinearRegression
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.types.StructType
import org.apache.spark.ml.evaluation.{BinaryClassificationEvaluator, Evaluator, MulticlassClassificationEvaluator, RegressionEvaluator}
import org.apache.spark.ml.feature.{Instance, LabeledPoint}

def generateLogisticInput(offset: Double,scale: Double,nPoints: Int,seed: Int):Seq[LabeledPoint] = {
    val rnd = new java.util.Random(seed)
    val x1 = Array.fill[Double](nPoints)(rnd.nextGaussian())
    val y = (0 until nPoints).map { i =>
      val p = 1.0 / (1.0 + math.exp(-(offset + scale * x1(i))))
      if (rnd.nextDouble() < p) 1.0 else 0.0
    }
    val testData = (0 until nPoints).map(i => LabeledPoint(y(i), Vectors.dense(Array(x1(i)))))
    testData
  }
import spark.implicits._
val dataset = sc.parallelize(generateLogisticInput(0.0, 1.0, 10, 42), 2).toDF()
val lr = new LogisticRegression
val lrParamMaps = new ParamGridBuilder().addGrid(lr.regParam, Array(0.001, 1000.0)).addGrid(lr.maxIter, Array(0, 3)).build()
val eval = new BinaryClassificationEvaluator
val numFolds = 3
val cv = new CrossValidator().setEstimator(lr).setEstimatorParamMaps(lrParamMaps).setEvaluator(eval).setNumFolds(numFolds)
val cvModel = cv.fit(dataset)
cvModel.save("file:///Users/weichenxu/work/test/s1")
```

and then run following code on current PR: (in spark-shell)
```
val model = org.apache.spark.ml.tuning.CrossValidatorModel.load("file:///Users/weichenxu/work/test/s1")
model.hasSubModels
```

SparkQA commented:
**[Test build #83824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83824/testReport)** for PR 19208 at commit [`7e997da`](https://github.com/apache/spark/commit/7e997da44157a9807de9c8fe8e7d2e5b66b6bfb1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83824/
Test FAILed.

WeichenXu123 commented:
Jenkins, test this please.

SparkQA commented:
**[Test build #83834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83834/testReport)** for PR 19208 at commit [`7e997da`](https://github.com/apache/spark/commit/7e997da44157a9807de9c8fe8e7d2e5b66b6bfb1).

SparkQA commented:
**[Test build #83835 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83835/testReport)** for PR 19208 at commit [`7e997da`](https://github.com/apache/spark/commit/7e997da44157a9807de9c8fe8e7d2e5b66b6bfb1).

SparkQA commented:
**[Test build #83835 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83835/testReport)** for PR 19208 at commit [`7e997da`](https://github.com/apache/spark/commit/7e997da44157a9807de9c8fe8e7d2e5b66b6bfb1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83835/
Test FAILed.

SparkQA commented:
**[Test build #83834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83834/testReport)** for PR 19208 at commit [`7e997da`](https://github.com/apache/spark/commit/7e997da44157a9807de9c8fe8e7d2e5b66b6bfb1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83834/
Test PASSed.

jkbradley commented:
Awesome, thanks for the updates and for checking backwards compatibility!
LGTM
Merging with master
